src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp

```diff
@@ -349,18 +382,18 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::WriteAttribute(const Dat
         }
     }
 
+    ContextAttributesChangeListener change_listener(CurrentContext());
+
     AttributeAccessInterface * aai =
         AttributeAccessInterfaceRegistry::Instance().Get(request.path.mEndpointId, request.path.mClusterId);
     std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
     if (aai_result.has_value())
     {
         if (*aai_result == CHIP_NO_ERROR)
         {
-            // TODO: change callbacks should likely be routed through the context `MarkDirty` only
-            //       however for now this is called directly because ember code does this call
-            //       inside emberAfWriteAttribute.
-            MatterReportingAttributeChangeCallback(request.path);
-            CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+            // TODO: this is awkward since it provides AAI no control over this, specifically
+            //       AAI may not want to increase versions for some attributes that are Q
+            emberAfAttributeChanged(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId, &change_listener);
```

Comment:
It's really odd to have any ember* thing in this code at all.  Why is it here?

Reply:
We need to do the same thing that ember does. We currently do a `MatterReportingAttributeChangeCallback` in https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L736 however that is not something this code does because we are supposed to use the dirty marking in the context rather than the global that `MatterReportingAttributeChangeCallback` is using.

Reply:
But the data version bump also needs to happen in the actual provider bits, not the ember data, no?

Reply:
Ah, this is the Codegen provider, so it's tied to ember, ok.